docs/specgen.md

```diff
@@ -0,0 +1,94 @@
+while supporting build tools and language servers with less effort and time.
+
+## Background
+
+The Build Server Protocol takes inspiration from the Language Server Protocol
+(LSP). Unlike in the Language Server Protocol, the language server or IDE is
+referred to as the “client” and a build tool such as sbt/Gradle/Bazel is
+referred to as the “server”.
+
+The best way to read this document is by considering it as a wishlist from the
+perspective of an IDE developer.
+
+The code listings in this document are written using TypeScript syntax. Every
+data strucuture in this document has a direct translation to JSON and Protobuf.
+
+## Relationship with LSP
+
+BSP can be used together with LSP in the same architecture. The diagram below
+illustrates an example how an LSP server can also act as a BSP client.
+
+![](https://i.imgur.com/q4KEas9.png)
+
+BSP can also be used without LSP. In the example above, IntelliJ acts as a BSP
+client even if IntelliJ does not use LSP.
+
+## Status
+
+The Build Server Protocol is not an approved standard. Everything in this
+document is subject to change and open for discussions, including core data
+structures.
+
+The creation of BSP clients and servers is under active development.
+
+In the clients space, IntelliJ has been the first language server to implement
+BSP. The integration is available in the nightly releases of the Scala plugin.
+Other language servers, like [Dotty IDE](https://github.com/lampepfl/dotty) and
+[scalameta/metals](https://github.com/scalameta/metals), are currently working
+or planning to work on a BSP integrations.
+
+On the server side,
+* [Bloop](https://github.com/scalacenter/bloop) was the first
+server to implement BSP
+* sbt added built-in support in [1.4.0](https://github.com/sbt/sbt/pull/5538),
+* Mill ships with [built-in BSP support](https://com-lihaoyi.github.io/mill/mill/Intro_to_Mill.html#_build_server_protocol_bsp)
+* Bazel support is provided by [bazel-bsp](https://github.com/JetBrains/bazel-bsp)
+
+We're looking for third parties that implement BSP natively in other build tools
+such as Gradle.
+
+The Build Server Protocol has been designed to be language-agnostic. We're
+looking for ways to collaborate with other programming language communities and
+build tool authors.
+
+The best way to share your thoughts on the Build Server Protocol or to get
+involved in its development is to open an issue or pull request to this
+repository. Any help on developing integrations will be much appreciated.
+
+## Base protocol
+
+The base protocol is identical to the language server base protocol. See
+<https://microsoft.github.io/language-server-protocol/specification> for more
+details.
+
+Like the language server protocol, the build server protocol defines a set of
+JSON-RPC request, response and notification messages which are exchanged using
+the base protocol.
+
+```scala mdoc:passthrough
+bsp.codegen.Codegen.printDocs()
```

Comment:
This wires the code-generation of docs to the markdown. The structuring of the spec changes a little to avoid doing convoluted shoehorning in this unit of work, but the heavylifting is mostly done.